// expire/lib/expire.js

var units = module.exports.units = {
    year:    31536000
  , month:   2592000
  , week:    604800
  , day:     86400
  , hour:    3600
  , minute:  60
  , second:  1
};

var values = module.exports.values = [
    'zero'
  , 'one'
  , 'two'
  , 'three'
  , 'four'
  , 'five'
  , 'six'
  , 'seven'
  , 'eight'
  , 'nine'
  , 'ten'
  , 'eleven'
  , 'twelve'
  , 'thirteen'
  , 'fourteen'
  , 'fifteen'
  , 'sixteen'
  , 'seventeen'
  , 'eighteen'
  , 'nineteen'
  , 'twenty'
].map(function(value, index) {
  return [value, ''+index];
});

var getValue = function getValue(v) {
  var res = 0;
  for (var i=0, len=values.length; i<len; i++) {
    if (values[i].indexOf(v) !== -1)
      return i;
  };
  return res;
};

var computeSeconds = function computeSeconds(value, unit) {
  unit = unit.replace(/s$/i, '');
  unit = parseInt(units[unit.toLowerCase()]);
  value = parseInt(getValue(value.toLowerCase()));
  if (!value || !unit) return 0;
  return value * unit;
};

module.exports.getSeconds = function getSeconds(str) {
  if (!str || typeof(str) !== 'string')
    return 0;

  var space = function(i) {
    return i.split(' ');
  };

  var subs = str.split(/\sand\s/).map(space);
  var sum = 0;

  subs.forEach(function(sub) {
    sum += computeSeconds.apply(this, sub)
  });

  return sum;
};

module.exports.setExpiration = function setExpiration(res, time) {
  if (!time) {
    time = res; res = {};
  };

  if (typeof(time) !== 'number') {
    time = module.exports.getSeconds(time);
  };

  res['Cache-Control'] = 'public, max-age='+time;
  return res;
};
